org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/cvc_complex_type_2_4_bCodeAction.java

```diff
@@ -0,0 +1,116 @@
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.dom.DOMElement;
+import org.eclipse.lemminx.dom.DOMNode;
+import org.eclipse.lemminx.extensions.contentmodel.model.CMDocument;
+import org.eclipse.lemminx.extensions.contentmodel.model.CMElementDeclaration;
+import org.eclipse.lemminx.extensions.contentmodel.model.ContentModelManager;
+import org.eclipse.lemminx.services.extensions.codeaction.ICodeActionParticipant;
+import org.eclipse.lemminx.services.extensions.codeaction.ICodeActionRequest;
+import org.eclipse.lemminx.extensions.contentmodel.utils.XMLGenerator;
+
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+import org.eclipse.lsp4j.Position;
+
+/**
+ * CodeAction to insert expected child elements within an element
+ * 
+ * Given this XML where the expected child elements are not present as defined
+ * in the XSD:
+ * 
+ * <servlet></servlet>
+ * Error:
+ * Child elements are missing from element:
+ * - servlet
+ *
+ * The following elements are expected:
+ * - description
+ * - display-name
+ * - icon
+ * - servlet-name
+ * 
+ * To fix the error, the code action will suggest inserting the expected
+ * elements inside the parent tag
+ * 
+ */
+
+public class cvc_complex_type_2_4_bCodeAction implements ICodeActionParticipant {
+
+	@Override
+	public void doCodeAction(ICodeActionRequest request, List<CodeAction> codeActions, CancelChecker cancelChecker) {
+		DOMDocument document = request.getDocument();
+		Diagnostic diagnostic = request.getDiagnostic();
+		try {
+			Range diagnosticRange = diagnostic.getRange();
+			int startOffset = document.offsetAt(diagnosticRange.getStart()) + 1;
+			DOMNode node = document.findNodeAt(startOffset);
+
+			if (node == null || !node.isElement()) {
+				return;
+			}
+
+			DOMElement element = (DOMElement) node;
+			int elementOffset = element.getEndTagOpenOffset();
+			Position childElementPosition = document.positionAt(elementOffset);
+
+			StringBuilder insertText = new StringBuilder();
+
+			XMLGenerator generator = request.getXMLGenerator();
+			ContentModelManager contentModelManager = request.getComponent(ContentModelManager.class);
+
+			// get list of all existing elements
+			Collection<String> existingElements = new TreeSet<String>(Collator.getInstance());
+			for (DOMNode root : document.getRoots()) {
+				existingElements.add(root.getNodeName());
+				Collection<DOMNode> existingChilren = root.getChildren();
+				for (DOMNode existingChild : existingChilren) {
+					existingElements.add(existingChild.getNodeName());
+				}
+			}
+
+			for (CMDocument cmDocument : contentModelManager.findCMDocument(element)) {
+				CMElementDeclaration matchesCMElement = cmDocument.findCMElement(element);
+				if (matchesCMElement != null) {
+					insertText.append(generator.generate(matchesCMElement, null, true, true, existingElements));
```

Comment:
I think it is not XML generator which must check that, but you must check that here:

`if (matchesCMElement != null && !existingElements.contains(matchesCMElement .getName())) {`

Reply:
I check in the XML generator because the `matchesCMElement` that gets passed in to `generate()` is the parent ('root'), and we need to check it against the children of root, which is extracted through an iterator inside the generate function. If we want to do the checks outside the generator, I believe we would need to bring the iterator outside and set `generateOnlyChildren` to false (or just remove that option).

Reply:
Thanks for your explanation, you are right. The check must be done in XMLGenerator but the existingElements should be filled like I suggested (just for the first level). The check with existingElements must be done only for the first level and nor for children.